core: vm: sqlvm: README.md (#294)
dexon-foundation/dexon: core/vm/sqlvm/README.md

CODE_SECTION = RLPEncode(Schema)
```

#### Primary index & data

##### List of keys

Key
```
PathKey(["tables", uint8(table_idx), "primary"])
```

Value
```golang
type PrimaryKeys struct {
	// Slot 1: Header.
	LastRowID   uint64
	RowCount    uint64
	<unused>    uint64
	<unused>    uint64
	// End of header.
	Slot2       uint256 // Bitmap, record 0 ~ 255.
	Slot3       uint256 // Bitmap, record 256 ~ 511.
	Slot4       uint256 // Bitmap, record 512 ~ 767.
    ...
}
```

Bitmap:

*   Use `uint64` to store primary index.
*   Pagination: 1 slot = 1 page
*   Reserve 1st slot for some information:
    *   Used slot number (0-63 bit)
    *   Record count (64-127 bit)
*   Not reuse bits
    * Only last bit will be reused.

##### Actual data
Key
```
PathKey(["tables", uint8(table_idx), "primary", uint64({row_id})])
```

Value
```
PackingEncode([column_1, column_2, column_3, ...])
```
`PackingEncode` will pack data if they can be put in one slot. It stores data slot by slot.
Dynamic bytes follow Solidity ABI encoding.

#### Other indices
##### List of keys
Key
```
PathKey(["tables", uint8(table_idx), "indices", uint8(index_idx)])
```

Value
```golang

> type PossibleValues struct {

Reviewer: ```suggestion
type IndexValues struct {
```

	// Slot 1.
	Length          uint64
	<unused>        uint64
	<unused>        uint64
	<unused>        uint64
	// Slot 2.
	ValueAddr1      uint256
	// Slot 3.
	ValueAddr2      uint256
	...
}
```
##### Actual data
Key
```
PathKey(["tables", uint8(table_idx), "indices", uint8(index_idx), field_1, field_2, field_3, ...])
```
Value
```golang

> type IndexKeys struct {

Reviewer: ```suggestion
type IndexEntry struct {
```

	// Slot 1.
	Length                      uint64

> 	IndexToIndexValuesOffset uint64

Reviewer: It is not aligned with other lines.

> 	IndexToPossibleValuesOffset uint64

Reviewer: ```suggestion
	IndexToIndexValuesOffset uint64
```

	ForeignKeyReferenceCount    uint64
	<unused>                    uint64
	// Slot 2.
	RowID1                      uint64
	RowID2                      uint64
	RowID3                      uint64
	RowID4                      uint64
	// Slot 3.
	RowID5                      uint64
	...
}
```

> 

Reviewer: "RowID reverse index" is missing.

Reviewer: It does not appear in the table of content.

Author: Added.

#### Sequence(auto increment)
Key
```
PathKey(["tables", uint8(table_idx), "sequence", uint8(sequence_idx)])
```

Value
```
uint64
```

#### NULL value
Do not support NULL value.

#### Compound index
Only (a, b, c) will be indexed.
(a), (b) or (a, b) won't be indexed automatically.

#### Contract owner
Key
```
PathKey(["owner"])
```

Value
```
address
```

#### Table writer

##### List of keys

Key
```
PathKey(["tables", uint8(table_idx), "writers"])
```

Value
```golang

> type PossibleValues struct {

Reviewer: Since the above `PossibleValues` has been renamed to `IndexValues` in our internal document, do you think we should rename this as well?

Reviewer: Do you really want to call it `IndexValues`? It is not used as a part of an index.

Author: Renamed to `Writers`.

Reviewer: It still looks inconsistent because we now have `Writers` and `TableWriter`, but let's ignore it and resolve naming issues later.

	// Slot 1.
	Length          uint64
	<unused>        uint64
	<unused>        uint64
	<unused>        uint64
	// Slot 2.
	<unused>        uint96
	Addr1           uint160
	// Slot 3.
	<unused>        uint96
	Addr2           uint160
	...
}
```

##### Actual data
Key
```

> PathKey(["tables", uint8(table_idx), "writers", "addr"])

Reviewer: ```suggestion
PathKey(["tables", uint8(table_idx), "writers", uint160(addr)])
```

```
Value
```golang
type TableWriter struct {
	// Slot 1.

> 	IndexToPossibleValuesOffset    uint64

Reviewer: This should be renamed if the above `PossibleValues` struct is renamed.

	// LT(t1, t2) res
	// res = t1 < t2 = [1, 2] < [2, 3] = [true, true]
	GT
	// GT(t1, t2) res
	// res = t1 > t2 = [1, 2] > [2, 3] = [false, false]
	EQ
	// EQ(t1, t2) res
	// res = t1 == t2 = [1, 2] == [2, 3] = [false, false]
	AND
	// AND(t1, t2) res
	// res = t1 && t2  = [true, true] && [true, false] = [true, false]
	OR
	// OR(t1, t2) res
	// res = t1 || t2 = [false, false] || [true, false] = [true, false]
	NOT
	// NOT(t1) res
	// res = !t1 = ![false, true] = [true, false]
	UNION
	// UNION(t1, t2) res
	// res = t1 ∪ t2  = [1, 2] ∪ [2, 3] = [1, 2, 3]
	INTXN
	// INTXN(t1, t2) res
	// res = t1 ∩ t2 = [1, 2] ∩ [2, 3] = [2]
	LIKE
	// LIKE(t1, pattern[, escape]) res
	// Immediate case
	// res = Like(t1, '%abc%') =
	//       ['_abc_', '123'] like ['%abc%'] = [true, false]
	// Not immediate case
	// res = Like(t1, t2, t3) =
	//      ['_abc', '1%23'] like ['%abc%', '_\%2_'] escape ["", "\"] =
	//      [true, true]
)

// 0x30 range - pk/index/field meta ops
const (
	REPEATPK = iota + 0x30
	// REPEATPK(table_id) res
	// res = [id1, id2, id3, ...]
	// REPEATPK(table_id) = [1, 2, 3, 5, 6, 7, ...]
	REPEATIDX
	// Scan given index value(s)
	// REPEATIDX(table_id, name_idx, idxv) res

> 	// res = [id2, id4, id5, id6]

Reviewer: The syntax looks inconsistent with the example. The syntax shows there are 3 parameters, while the example only has 2.

	// REPEATIDX(
	//  [table_id, name_idx],
	//  [val1, val3]
	// ) = [5, 6, 7, 10, 11, ... ]
	REPEATIDXV
	// Get possible values from index value meta
	// REPEATIDXV(table_id, name_idx) res

> 	// res = [val1, val2, val3, ...]

Reviewer: This is again inconsistent. The syntax has 2 parameters, but there is only 1 argument in the example.

	//   [table_id, name_idx]
	// ) = ["alice", "bob", "foo", "bar", ... ]
)

// 0x40 range - format/output ops
const (
	ZIP = iota + 0x40
	// ZIP(tgr, new) = res
	// ZIP([f1, f2, f3], [c1, c2, c3]) = [(f1, c1), (f2, c2), (f3, c3)]
	// ZIP(
	//    [(f1, c1), (f2, c2), (f3, c3)],
	//    [(x1, y1), (x2, y2), (x3, y3)]
	// ) = [(f1, c1, x1, y1), (f2, c2, x2, y2), (f3, c3, x3, y3)]
	FILED
	// FIELD(src, fields) = res
	// FIELD(
	//    [(r1f0, r1f1, r1f2), (r2f0, r2f1, r2f2),...], [(1, 2)]
	// ) = [(r1f1, r1f2), (r2f1, r2f2), ...]
	PRUNE // in-place op
	// PRUNE(src, fields) = res
	// PRUNE(
	//     [(r1f0, r1f1, r1f2), (r2f0, r2f1, r2f2),...], [1]
	// ) = [(r1f0, r1f2), (r2f0, r2f2), ...]
	SORT // in-place op
	// SORT(src, (ascending bool, field idx))
	// SORT(src, [(ascending, field)] ) = res
	// SORT(
	//   [(a1, a2), (b1, a2), (a1, b2), ...],
	//   [(asc, 0), (desc, 1), (asc, 2)]
	// ) = [(a1, a2), (a1, b2), (b1, a2), ...]
	FILTER
	// FILTER(src, cond) = res
	// FILTER([1, 2, 3, 4, 5], [true, false, true, false, false]) = [1, 3]
	CAST
	// CAST(t1, type) t2
	CUT
	// in-place op
	// CUT(src, range) = res
	// CUT(src, (start[, end]))
	// CUT(
	//      [(r1f1, r1f2, r1f3, r1f4),

> 	//      (r2f1, r2f2, r2f3, r2f4), ...], [(1, 2)]

Reviewer: ```suggestion
	//       (r2f1, r2f2, r2f3, r2f4), ...], [(1, 2)]
```

)

// 0x50 range - function ops
const (
	FUNC = iota + 0x50
	// FUNC(t1, func id[, args...])) = res
)

// 0x60 range - storage ops
const (
	INSERT = iota + 0x60
	// INSERT(table_id, fields, values...) = res
	// the number of values depends on the length of fields
	// res not important
	// INSERT(
	//    table_id,
	//    [0, 2, 5],
	//    [field0],
	//    [field2],
	//    [Immediate value],
	// ) = _
	UPDATE
	// UPDATE(table_id, ids, fields, values...) = res
	// the number of values depends on the length of fields
	// res = uint64(affected row)
	// UPDATE(
	//    table_id,
	//    [0, 99],
	//    [1, 2, 3],
	//    [updated_field0-1, updated_field99-1],
	//    [updated_field0-2, updated_field99-2],
	//    [Immediate Value],
	// ) = uint64(2)
	LOAD
	// LOAD(table_id, ids, fields) = res
	// LOAD(
	//   table_id,
	//   [55, 66],
	//   [1, 2, 3],
	// ) = [
	//   (field55-1, field55-2, field55-3),
	//   (field66-1, field66-2, field66-3),
	// ]
	DELETE
	// DELETE(table_id, ids) = res
	// res = uint64(affected row)
	// DELETE(
	//    table_id,
	//    [1, 2, 3, 100],
	// ) = uint64(4)
)
```

##### Examples
*   SELECT * FROM a
    *   REPEATPK(a) t1

>     *   LOAD(a, t1, [0,1,2,3,4,5]) t2

Reviewer: ```suggestion
    *   LOAD(a, t1, [0, 1, 2, 3, 4, 5]) t2
```

>     *   ZIP(t2)

Reviewer: What does `ZIP` do here? It looks unnecessary.

*   SELECT f1, f3, f5 FROM a WHERE f1=val (index hit)
    *   REPEATIDXV(a_f1_idx) t1

>     *   EQ(t1, val) t3

Reviewer: Is the name `t2` skipped deliberately?

    *   FILTER(t1, t3) t4
    *   REPEATIDX(a_f1_idx, t4) t5

>     *   LOAD(a, t5, [1,3,5]) t6

Reviewer: ```suggestion
    *   LOAD(a, t5, [1, 3, 5]) t6
```

    *   ZIP(t6)
*   SELECT f1, f3, f5 FROM a WHERE f1 > val (index scan)
    *   REPEATIDXV(a_f1_idx) t1
    *   GT(t1, val) t2
    *   FILTER(t1, t2) t3
    *   REPEATIDX(a_f1_idx, t3) t4
    *   LOAD(a, t4, [1, 3, 5]) t5
    *   ZIP(t5)
*   SELECT f1 > f2, f5 FROM a WHERE f3 > 10 (field operation)
    *   REPEATPK(a) t1
    *   LOAD(a, t1, [3]) t2
    *   GT(t2, 10) t4
    *   FILTER(t1, t4) t5
    *   LOAD(a, t5, [1, 2, 5]) t2
    *   FIELD(t2, [1]) t3
    *   FIELD(t2, [2]) t4
    *   GT(t3, t4) t5
    *   FIELD(t2, [5]) t9
    *   ZIP(t5, t9) t10

> *   DELETE FROM a WHERE f2 = 'c' (idx hit)

Reviewer: ```suggestion
*   DELETE FROM a WHERE f2 = 'c' (index hit)
```

    *   REPEATIDX(a_f2_idx, ['c']) t1
    *   DELETE(a, t1)
*   UPDATE a SET f1 = 'new'
    *   REPEATPK(a) t1
    *   UPDATE(a, t1, [1], ['new'])
*   INSERT a VALUES (f1, f2, f3), (g1, g2, g3);

>     *   INSERT(a, [1,2,3], (f1, f2, f3))

Reviewer: ```suggestion
    *   INSERT(a, [1, 2, 3], (f1, f2, f3))
```

>     *   INSERT(a, [1,2,3], (g1, g2, g3))

Reviewer: ```suggestion
    *   INSERT(a, [1, 2, 3], (g1, g2, g3))
```

> *   INSERT INTO a (c0,c3,c5) VALUES (NOW(), MSGSENDER(), RAND());

Reviewer: ```suggestion
*   INSERT INTO a (c0, c3, c5) VALUES (NOW(), MSGSENDER(), RAND());
```

    *   FUNC('NOW') t1          // t1 = [(block_height)]
    *   FUNC('MEGSENDER') t2    // t2 = [(address     )]
    *   FUNC('RAND') t3         // t3 = [(rand_num    )]
    *   ZIP(t1, t2) t1          // t1 = [(block_height, address)]
    *   ZIP(t1, t3) t1          // t1 = [(block_height, address, rand_num)]

>     *   INSERT(a, [0,3,5], t1)

Reviewer: ```suggestion
    *   INSERT(a, [0, 3, 5], t1)
```

1. `select random() from table order by random();` (in sqlite, it is equivalent to 'order by 1')
   1. those two random() are different.
1. `insert into table (a) values (random()+1), (1), (random());` (expression in insert)
1. `select * from table where column > random();` (we cannot use index on column as random is in condition)

## **Miscellaneous**

* If there is constraint conflict, abort transaction.
* (TBD) There is a default return limit for avoiding out of gas.
* There is a flag to decide to enable flow check or not.
* It can not transfer token to SQL contract address.
* Do not support expression default value.
* AUTOINCREMENT increases in the max value, can't auto increment with foreign key, only int/uint can use auto increment, dominate over 'default' setting `
* AUTOINCREMENT starts from 1.
* Unique check (in memory check, i.e. rehearsal in memory)`
* Dynamic Bytes support:
    * `OCTET_LENGTH`
    * `SUBSTRING`
* Use bytes to store fixed bytes data
* Fixed bytes supports:
    * `OCTET_LENGTH`
    * `fixBytes3Data || fixBytes5Data`
        * Return bytes8
        * Abort if larger than 32

>     * CAST fixBytes to DynamciByte and Dynamicbyte to fixBytes

Reviewer: ```suggestion
    * CAST fixed bytes to dynamic bytes and dynamic bytes to fixed bytes
```
